README.md
# clj-challenges

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-03-31
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-03-31
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clj-challenges/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clj-challenges/compare/0.1.0...0.1.1


test/clj_challenges/core_test.clj
(ns clj-challenges.core-test
  (:require [clojure.test :refer [deftest is testing]]
            [clj-challenges.core :refer [foo]]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 1 1))))


test/clj_challenges/4clojure_test.clj
(ns clj-challenges.4clojure-test
  (:require [clojure.test :refer [deftest is testing]]
            [clj-challenges.4clojure]))

(deftest nothing_but_the_truth_test (testing "https://4clojure.oxal.org/#/problem/1"
    (is (true? (clj-challenges.4clojure/nothing_but_the_truth)))))

(deftest simple_math_test (testing "https://4clojure.oxal.org/#/problem/2"
    (is (true? (clj-challenges.4clojure/simple_math)))))

(deftest strings_test (testing "https://4clojure.oxal.org/#/problem/3"
    (is (true? (clj-challenges.4clojure/strings)))))

(deftest lists_test (testing "https://4clojure.oxal.org/#/problem/4"
    (is (true? (clj-challenges.4clojure/lists)))))

(deftest conj_on_lists_test (testing "https://4clojure.oxal.org/#/problem/5"
    (is (true? (clj-challenges.4clojure/conj_on_lists)))))

(deftest vectors_test (testing "https://4clojure.oxal.org/#/problem/6"
    (is (true? (clj-challenges.4clojure/vectors)))))

(deftest conj_on_vectors_test (testing "https://4clojure.oxal.org/#/problem/7"
    (is (true? (clj-challenges.4clojure/conj_on_vectors)))))

(deftest sets_test (testing "https://4clojure.oxal.org/#/problem/8"
    (is (true? (clj-challenges.4clojure/sets)))))

(deftest conj_on_sets_test (testing "https://4clojure.oxal.org/#/problem/9"
    (is (true? (clj-challenges.4clojure/conj_on_sets)))))

(deftest maps_test (testing "https://4clojure.oxal.org/#/problem/10"
    (is (true? (clj-challenges.4clojure/maps)))))

(deftest conj_on_maps_test (testing "https://4clojure.oxal.org/#/problem/11"
    (is (true? (clj-challenges.4clojure/conj_on_maps)))))

(deftest sequences_test (testing "https://4clojure.oxal.org/#/problem/12"
    (is (true? (clj-challenges.4clojure/sequences)))))

(deftest rest_test (testing "https://4clojure.oxal.org/#/problem/13"
    (is (true? (clj-challenges.4clojure/clj-rest)))))

(deftest functions_test (testing "https://4clojure.oxal.org/#/problem/14"
    (is (true? (clj-challenges.4clojure/functions)))))


test/clj_challenges/first_factorial_test.clj
(ns clj-challenges.first-factorial-test
  (:require [clojure.test :refer [deftest is testing]]
            [clj-challenges.first-factorial]))

(deftest zero_is_zero
  (testing "zero gives result zero"
    (is (= 0 (clj-challenges.first-factorial/FirstFactorial 0)))))

(deftest test_nil_returnsZero
  (testing "nil input gives zero as output"
    (is (= 0 (clj-challenges.first-factorial/FirstFactorial nil)))))

(deftest test_one_returns_one
  (testing "1 as input gives 1 as output"
    (is (= 1 (clj-challenges.first-factorial/FirstFactorial 1)))))

(deftest test_two_returns_2
  (testing "2 as input gives 2 as output"
    (is (= 2 (clj-challenges.first-factorial/FirstFactorial 2)))))

(deftest test_not_a_number_gives_output_zero
  (testing
    (is (= 0 (clj-challenges.first-factorial/FirstFactorial "no strings are allowed")))))

(deftest test_that_higher_numbers_do_the_factorial
  (testing
    (is (= (* 1 2 3 4) (clj-challenges.first-factorial/FirstFactorial 4)))))


src/clj_challenges/4clojure.clj
(ns clj-challenges.4clojure
  (:require [clojure.set]))

(defn nothing_but_the_truth "https://4clojure.oxal.org/#/problem/1" []
  (= true true))

(defn simple_math "https://4clojure.oxal.org/#/problem/2" []
  (= (- 10 (* 2 3)) 4))

(defn strings "https://4clojure.oxal.org/#/problem/3" []
  (= "HELLO WORLD" (.toUpperCase "hello world")))

(defn lists "https://4clojure.oxal.org/#/problem/4" []
  (= (list :a :b :c) '(:a :b :c)))

(defn conj_on_lists "https://4clojure.oxal.org/#/problem/5" []
  (and (= '(1 2 3 4) (conj '(2 3 4) 1)) 
       (= '(1 2 3 4) (conj '(3 4) 2 1))))

(defn vectors "https://4clojure.oxal.org/#/problem/6" []
  (= [:a :b :c] (list :a :b :c) (vec '(:a :b :c)) (vector :a :b :c)))

(defn conj_on_vectors "https://4clojure.oxal.org/#/problem/7" []
  (let [result [1 2 3 4]]
  (and (= result (conj [1 2 3] 4))
       (= result (conj [1 2] 3 4))))) 

(defn sets "https://4clojure.oxal.org/#/problem/8" []
  (let [answer #{:a :b :c :d}]
    (and (= answer (set '(:a :a :b :c :c :c :c :d :d))) 
         (= answer (clojure.set/union #{:a :b :c} #{:b :c :d})))))

(defn conj_on_sets "https://4clojure.oxal.org/#/problem/9" []
  (= #{1 2 3 4} (conj #{1 4 3} 2)))

(defn maps "https://4clojure.oxal.org/#/problem/10" []
  (let [answer 20]
    (and (= answer ((hash-map :a 10, :b 20, :c 30) :b)) 
         (= answer (:b {:a 10, :b 20, :c 30})))))

(defn conj_on_maps "https://4clojure.oxal.org/#/problem/11" []
  (= {:a 1, :b 2, :c 3} (conj {:a 1} {:b 2} [:c 3])))

(defn sequences "https://4clojure.oxal.org/#/problem/12" []
  (let [__ 3]
    (and (= __ (first '(3 2 1))) 
         (= __ (second [2 3 4]))
         (= __ (last (list 1 2 3))))))

(defn clj-rest "https://4clojure.oxal.org/#/problem/13" []
  (= '(20 30 40) (rest [10 20 30 40])))

(defn functions "https://4clojure.oxal.org/#/problem/14" []
  (let [__ 8]
    (and (= __ ((fn add-five [x] (+ x 5)) 3)) 
         (= __ ((fn [x] (+ x 5)) 3))
         (= __ (#(+ % 5) 3))
         (= __ ((partial + 5) 3)))))
